### PR TITLE
WIP:Blstm

### PR DESCRIPTION
**Pull request by freewym**
Need to test the back-compatible.
I Didn't explictly add an option for choosing blstm or blstm. Instead, BLSTM layer would be recognized from the option "--lstm-delay", eg, if lstm_delay=" -1,1 -2,2 -3,3 " will be parsed as 3 blstm layers with delays of 1, 2, 3

### CODE REVIEW

**Comment by danpovey** [comment]
Thanks!

Of course, this patch will conflict with @naxingyu's patch.   My working
assumption is that we will take @naxingyu's patch because he was ready
first, and he has more results; and we may just take results from you (if
it turns out your setups are equivalent).  But I will leave the final
decision (and any review) to @vijayaditya.

**Comment by freewym** [comment]
OK. I am taking a look at xingyu's.

**Comment by vijayaditya** [comment]
I would recommend changing the format of the lstm_delay string. I think 

e.g. "-1 [-1,1] -2 [-2,2]" would mean lstm with delay 1 + blstm with delay 1 + lstm with delay 2 + blstm with delay 2. 

I recommend this format as it obvious from a brief glance whether an arch has BLSTM nodes or not, because of the square brackets.

File: egs/wsj/s5/steps/nnet3/lstm/decode.sh [hunk 1]
```diff
@@ -30,6 +30,8 @@ minimize=false
 frames_per_chunk=10000
 extra_left_context=20   # it is recommended to use the same value as the chunk_left_context
                         # used during training
+extra_right_context=20  # it is recommended to use the same value as the chunk_right_context
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/decode.sh]: the default value has to be null string or 0, as the default recipe is lstm unless specified.

**Comment by danpovey** [comment]
Vijay, before asking for changes in yiming's patch I think you need to
decide whether we will go with his patch or Xingyu's.

Dan

File: egs/wsj/s5/steps/nnet3/lstm/make_configs.py [hunk 2]
```diff
@@ -59,6 +59,24 @@ def ParseSpliceString(splice_indexes, label_delay=None):
             'num_hidden_layers':len(splice_array)
             }
 
+def ParseLstmDelayString(lstm_delay):
+    ## Work out lstm_delay e.g. lstm_delay = [ [-1], [-1, 1], [-2] ]
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/make_configs.py]: The lstm_delay format in the comment does not match your input. Make them consistent.

File: egs/wsj/s5/steps/nnet3/lstm/train.sh [hunk 3]
```diff
@@ -66,6 +66,7 @@ clipping_threshold=30     # if norm_based_clipping is true this would be the max
                           # else this is the max-absolute value of each element in Jacobian.
 chunk_width=20  # number of output labels in the sequence used to train an LSTM
 chunk_left_context=40  # number of steps used in the estimation of LSTM state before prediction of the first label
+chunk_right_context=40  # number of steps used in the estimation of LSTM state before prediction of the first label
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: default chunk_right_context has to be 0, as the default recipe is LSTM. If you want you can can run the make_configs.py script to figure out if the arch. has a BLSTM and use a new set of defaults in this case,.

**Comment by vijayaditya** [comment]
OK. Let me take a look.

File: egs/wsj/s5/local/nnet3/run_blstm.sh [hunk 4]
```diff
@@ -0,0 +1,128 @@
+#!/bin/bash
```
Comment (vijayaditya) [inline egs/wsj/s5/local/nnet3/run_blstm.sh]: I think you should eliminate this file, as we could just use lstm_delay string with the usual run_lstm.sh to specify blstms.

Comment (freewym) [inline egs/wsj/s5/local/nnet3/run_blstm.sh]: Yes agree. I will remove it eventually after I got the result of a good configuration of blstm.

Thanks,
Yiming

**Comment by freewym** [comment]
@vijayaditya Please check the fix. (I will remove run_blstm.sh eventually)

The 1st one is for LSTM as reference. Among the latter 4 BLSTM configurations, not a single one is consistently better than the others on all of the 4 datasets. Perhaps the last configuration is a little better, which has the closest num-parameters to the LSTM's.

LSTM cell_dim=1024, recurrent_projection_dim=non_recurrent_projection_dim=256,lstm_delay=-1 -2 -3, label_delay=5, num-parameters: 11894059
%WER 7.32 [ 603 / 8234, 70 ins, 92 del, 441 sub exp/nnet3/lstm_ld5/decode_bd_tgpr_dev93/wer_11_0.0
%WER 4.24 [ 239 / 5643, 22 ins, 23 del, 194 sub ] exp/nnet3/lstm_ld5/decode_bd_tgpr_eval92/wer_10_1.0
%WER 9.57 [ 788 / 8234, 151 ins, 95 del, 542 sub ] exp/nnet3/lstm_ld5/decode_tgpr_dev93/wer_9_1.0
%WER 6.86 [ 387 / 5643, 87 ins, 23 del, 277 sub ] exp/nnet3/lstm_ld5/decode_tgpr_eval92/wer_10_1.0

BLSTM cell_dim=1024, recurrent_projection_dim=non_recurrent_projection_dim=256,lstm_delay=[-1,1] [-2,2] [-3,3], label_delay=0, num-parameters: 32173355
%WER 7.07 [ 582 / 8234, 62 ins, 94 del, 426 sub ] exp/nnet3/blstm_1024_123_ld0/decode_bd_tgpr_dev93/wer_11_0.0
%WER 4.31 [ 243 / 5643, 19 ins, 31 del, 193 sub ] exp/nnet3/blstm_1024_123_ld0/decode_bd_tgpr_eval92/wer_11_1.0
%WER 9.67 [ 796 / 8234, 140 ins, 95 del, 561 sub ] exp/nnet3/blstm_1024_123_ld0/decode_tgpr_dev93/wer_12_0.0
%WER 6.57 [ 371 / 5643, 73 ins, 25 del, 273 sub ] exp/nnet3/blstm_1024_123_ld0/decode_tgpr_eval92/wer_13_0.5

BLSTM cell_dim=1024, recurrent_projection_dim=non_recurrent_projection_dim=128,lstm_delay=[-1,1] [-2,2] [-3,3], label_delay=0, num-parameters: 17338667
%WER 6.69 [ 551 / 8234, 52 ins, 94 del, 405 sub ] exp/nnet3/blstm_1024_123_ld0/decode_bd_tgpr_dev93/wer_12_0.0
%WER 4.38 [ 247 / 5643, 33 ins, 14 del, 200 sub ] exp/nnet3/blstm_1024_123_ld0/decode_bd_tgpr_eval92/wer_9_0.0
%WER 9.49 [ 781 / 8234, 124 ins, 99 del, 558 sub ] exp/nnet3/blstm_1024_123_ld0/decode_tgpr_dev93/wer_11_0.5
%WER 6.79 [ 383 / 5643, 86 ins, 17 del, 280 sub ] exp/nnet3/blstm_1024_123_ld0/decode_tgpr_eval92/wer_10_0.5

BLSTM cell_dim=768, recurrent_projection_dim=non_recurrent_projection_dim=128 lstm_dealy=[-1,1] [-2,2] [-3,3], label_delay=0, num-parameters: 13436715
%WER 7.21 [ 594 / 8234, 62 ins, 102 del, 430 sub ] exp/nnet3/blstm_ld0/decode_bd_tgpr_dev93/wer_13_0.0
%WER 4.24 [ 239 / 5643, 19 ins, 24 del, 196 sub ] exp/nnet3/blstm_ld0/decode_bd_tgpr_eval92/wer_10_1.0
%WER 9.62 [ 792 / 8234, 144 ins, 88 del, 560 sub ] exp/nnet3/blstm_ld0/decode_tgpr_dev93/wer_11_0.0
%WER 6.63 [ 374 / 5643, 88 ins, 16 del, 270 sub ] exp/nnet3/blstm_ld0/decode_tgpr_eval92/wer_9_0.5

BLSTM cell_dim=640, recurrent_projection_dim=non_recurrent_projection_dim=128,lstm_delay=[-1,1] [-2,2] [-3,3], label_delay=0, num-parameters: 11485739
%WER 6.81 [ 561 / 8234, 59 ins, 95 del, 407 sub ] exp/nnet3/blstm_640_123_ld0/decode_bd_tgpr_dev93/wer_11_0.0
%WER 4.27 [ 241 / 5643, 30 ins, 20 del, 191 sub ] exp/nnet3/blstm_640_123_ld0/decode_bd_tgpr_eval92/wer_11_0.0
%WER 9.29 [ 765 / 8234, 136 ins, 92 del, 537 sub ] exp/nnet3/blstm_640_123_ld0/decode_tgpr_dev93/wer_11_0.5
%WER 6.61 [ 373 / 5643, 71 ins, 28 del, 274 sub ] exp/nnet3/blstm_640_123_ld0/decode_tgpr_eval92/wer_11_1.0

**Comment by danpovey** [comment]
The WSJ test sets are very small by modern standards and you can't expect
the results on all 4 of those conditions to be consistent.  Just take the
average.

**Comment by freewym** [comment]
Thanks. So the the last BLSTM setting, with approximately the same num-params as LSTM's, reduced the avg WER from 7% to 6.745%.

Yiming

**Comment by vijayaditya** [comment]
I would suggest you run a test on a larger database like SWBD or a tougher task like AMI which has higher WER for baseline where relative WER changes are more noticeable.

**Comment by freewym** [comment]
Results On swbd data:

LSTMs cell_dim=1024, recurrent_dim=256, lstm_delay=-1 -2 -3, label_delay=5
%WER 15.55
%WER 16.62
%WER 11.6
%WER 13.1

BLSTMs cell_dim=1024, recurrent_dim=128, lstm_delay=-1 -2 -3, label_delay=0
%WER 13.55
%WER 14.33
%WER 10.3
%WER 11.3

BLSTMs cell_dim=1024, recurrent_dim=256, lstm_delay=-1 -2 -3, label_delay=0
%WER 15.83
%WER 16.70 
%WER 12.0 
%WER 13.1

**Comment by danpovey** [comment]
Interesting.  How many more parameters does the 'good' BLSTM have than the
LSTM?  Have we ever run the LSTM baseline with more parameters?
Dan

**Comment by freewym** [comment]
The num-params are added in the above post. I didn't run LSTM baseline with more parameters, but I ran BLSTM with cell_dim=640, recurrent_dim=128, lstm_delay=-1 -2 -3, label_delay=0, num_params=14248244, which has similar number of parameters to the basline's, but gave poorer train and cv accuracies than the basline during training. So I terminated that setting.

**Comment by danpovey** [comment]
OK.  The BLSTM has a really huge number of parameters: 38 million, versus
15 million for the LSTM baseline.  I think running the LSTM baseline with
more parameters, e.g. about 25 million, by increasing the cell-dim and
recurrent-dim appropriately, might tell us something.
Dan

**Comment by danpovey** [comment]
Also, please let me know the directory names.  I want to see the avg-deriv
to see if the shrinking is working.

**Comment by freewym** [comment]
I will run baseline with more parameters.

The directories are
/export/a12/ywang/kaldi/egs/swbd/s5c/exp/nnet3/
lstm_ld5_sp
blstm_1024_128_123_sp
blstm_1024_256_123_sp

**Comment by vijayaditya** [comment]
What do these individual WERs correspond to ? Usually it is helpful if you
paste the entire output of the command in RESULTS file, for your experiment.

--Vijay

**Comment by freewym** [comment]
I have updated that post.

Yiming

File: egs/ami/s5/local/nnet3/run_ivector_common.sh [hunk 5]
```diff
@@ -96,20 +96,20 @@ if [ $stage -le 5 ] && [ "$speed_perturb" == "true" ]; then
 
   mfccdir=mfcc_${mic}_perturbed
   for x in train_sp; do
-    steps/make_mfcc.sh --cmd "$train_cmd" --nj $nj \
+    steps/make_mfcc.sh --cmd "$train_cmd" --nj 10 \
```
Comment (vijayaditya) [inline egs/ami/s5/local/nnet3/run_ivector_common.sh]: why did you remove $nj ?

Comment (danpovey) [inline egs/ami/s5/local/nnet3/run_ivector_common.sh]: looks like $nj may not be defined in the script.

File: egs/swbd/s5c/RESULTS [hunk 6]
```diff
@@ -109,6 +109,20 @@ exit 0
 %WER 11.8 | 1831 21395 | 89.4 7.0 3.6 1.3 11.8 47.8 | exp/nnet3/lstm_ld5_sp/decode_eval2000_sw1_fsh_fg/score_13_0.0/eval2000_hires.ctm.swbd.filt.sys
 %WER 13.2 | 1831 21395 | 88.3 8.0 3.7 1.5 13.2 49.5 | exp/nnet3/lstm_ld5_sp/decode_eval2000_sw1_tg/score_11_0.0/eval2000_hires.ctm.swbd.filt.sys
 
+# these are results with nnet3 LSTMs cell_dim=1280, recurrent_dim=384, lstm_delay=-1 -2 -3, label_delay=5 num_params=25010228 (8 epoch training on speed-perturbed
+# and volume perturbed data)
+%WER 15.20 [ 7477 / 49204, 1031 ins, 1912 del, 4534 sub ] exp/nnet3/lstm_1280_384_123_ld5_sp/decode_train_dev_sw1_fsh_fg/wer_12_0.0
```
Comment (vijayaditya) [inline egs/swbd/s5c/RESULTS]: Just give results for eval set.
Replace the default parameter settings for the LSTM with the those that give you best result. Remove the old LSTM results. The RESULTS file should have results that can be reproduced by running the local/nnet3/run_lstm.sh script.

File: egs/swbd/s5c/RESULTS [hunk 7]
```diff
@@ -109,6 +109,20 @@ exit 0
...
+%WER 16.29 [ 8013 / 49204, 1121 ins, 1908 del, 4984 sub ] exp/nnet3/lstm_1280_384_123_ld5_sp/decode_train_dev_sw1_tg/wer_11_0.0
+%WER 11.4 | 1831 21395 | 89.8 6.8 3.4 1.2 11.4 46.0 | exp/nnet3/lstm_1280_384_123_ld5_sp/decode_eval2000_sw1_fsh_fg/score_11_0.0/eval2000_hires.ctm.swbd.filt.sys
+%WER 12.7 | 1831 21395 | 88.6 7.6 3.8 1.3 12.7 48.7 | exp/nnet3/lstm_1280_384_123_ld5_sp/decode_eval2000_sw1_tg/score_11_0.0/eval2000_hires.ctm.swbd.filt.sys
+
+# these are results with nnet3 bidirectional LSTMs cell_dim=1024, recurrent_dim=128, lstm_delay=[-1,1] [-2,2] [-3,3] num_params=20101172 (8 epoch training on
```
Comment (vijayaditya) [inline egs/swbd/s5c/RESULTS]: specify which script has to be run to generate these results.

File: egs/wsj/s5/local/nnet3/run_lstm.sh [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+recurrent_projection_dim=256
+non_recurrent_projection_dim=256
+chunk_width=20
+chunk_left_context=40
+chunk_right_context=0
+
+
+# training options
+num_epochs=10
+initial_effective_lrate=0.0006
```
Comment (vijayaditya) [inline egs/wsj/s5/local/nnet3/run_lstm.sh]: I think we use smaller learning rates in AMI and SWBD scripts. Is there a reason for choosing these learning rates?

File: egs/wsj/s5/steps/nnet3/lstm/make_configs.py [hunk 9]
```diff
@@ -59,6 +59,24 @@ def ParseSpliceString(splice_indexes, label_delay=None):
...
             }
 
+def ParseLstmDelayString(lstm_delay):
+    ## Work out lstm_delay e.g. "-1 [-1,1] -2" -> list([ [-1], [-1, 1], [-2] ])
+    split1 = lstm_delay.split(" ");
+    lstm_delay_array = []
+    try:
+        for i in range(len(split1)):
+            indexes = map(lambda x: int(x), split1[i].strip().lstrip('[').rstrip(']').strip().split(","))
+            print(indexes)
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/make_configs.py]: remove this print statement.

File: egs/wsj/s5/steps/nnet3/lstm/make_configs.py [hunk 10]
```diff
@@ -170,17 +188,42 @@ def ParseSpliceString(splice_indexes, label_delay=None):
...
-                                         args.clipping_threshold, args.norm_based_clipping,
-                                         args.ng_per_element_scale_options, args.ng_affine_options,
-                                         lstm_delay = lstm_delay[i])
+	if len(lstm_delay[i]) == 2: # BLSTM layer case, add both forward and backward
+	    try:
+	        if not (lstm_delay[i][0] < 0 and lstm_delay[i][1] > 0):
+	            raise ValueError('Warning: ' + str(lstm_delay[i]) + ' is not a standard BLSTM mode. There should be a negative delay for the forward, and a postive delay for the backward.')
+	    except ValueError as e:
+		print(str(e))
+            prev_layer_output1 = nodes.AddLstmLayer(config_lines, "Lstm{0}f".format(i+1), prev_layer_output, args.cell_dim,
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/make_configs.py]: We already use the name "f" for forget gates in the LSTM. I would suggest changing the layer names to BLSTM{0}_forward and  BLSTM{0}_backward.

File: egs/wsj/s5/steps/nnet3/lstm/make_configs.py [hunk 11]
```diff
@@ -59,6 +59,24 @@ def ParseSpliceString(splice_indexes, label_delay=None):
...
+    split1 = lstm_delay.split(" ");
+    lstm_delay_array = []
+    try:
+        for i in range(len(split1)):
+            indexes = map(lambda x: int(x), split1[i].strip().lstrip('[').rstrip(']').strip().split(","))
+            print(indexes)
+            if len(indexes) < 1:
+                raise ValueError("invalid --lstm-delay argument, too-short element: "
+                                + lstm_delay)
+            lstm_delay_array.append(indexes)
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/make_configs.py]: You should probably do a check to ensure that there are both negative and positive delays.

File: egs/wsj/s5/steps/nnet3/lstm/make_configs.py [hunk 12]
```diff
@@ -170,17 +188,42 @@ def ParseSpliceString(splice_indexes, label_delay=None):
...
+			                    args.recurrent_projection_dim, args.non_recurrent_projection_dim,
+					    args.clipping_threshold, args.norm_based_clipping,
+					    args.ng_per_element_scale_options, args.ng_affine_options,
+					    lstm_delay = lstm_delay[i][0])
         # make the intermediate config file for layerwise discriminative
         # training
         nodes.AddFinalLayer(config_lines, prev_layer_output, args.num_targets, args.ng_affine_options, args.label_delay)
         config_files['{0}/layer{1}.config'.format(args.config_dir, i+1)] = config_lines
         config_lines = {'components':[], 'component-nodes':[]}
+	if len(lstm_delay[i]) == 2:
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/make_configs.py]: Clearly describe why you are doing this descriptor modification. I remember that you needed it as you could cascade Append descriptors. It is better that you mention it here for the sake of clarity.

File: egs/swbd/s5c/RESULTS (hunk 7 above)
Comment (vijayaditya) [inline egs/swbd/s5c/RESULTS]: As you don't have a separate top level script for BLSTMs I would recommend that you modify 
https://github.com/kaldi-asr/kaldi/blob/master/egs/swbd/s5c/run.sh with the command which will run BLSTM

e.g.
# nnet3 BLSTM recipe
# local/nnet3/run_lstm.sh --has-fisher $has_fisher <required-arguments>

Also update the RESULTS file to specify that this command has to be run to generate the results. Do the same for WSJ too.

File: egs/ami/s5/local/nnet3/run_ivector_common.sh (hunk 5 above)
Comment (vijayaditya) [inline egs/ami/s5/local/nnet3/run_ivector_common.sh]: @freewym better add the nj variable. It makes life easy when you want to modify the number of jobs.

File: egs/swbd/s5c/RESULTS (hunk 7 above)
Comment (vijayaditya) [inline egs/swbd/s5c/RESULTS]: I see that the run.sh script does not have the commands for LSTM script. Could you please add a command for that too.
